src/lakefs_spec/util.py

```diff
@@ -1,11 +1,33 @@
 import hashlib
 import re
+from typing import Callable, Generator, ParamSpec, Protocol, TypeVar
 
 from lakefs_sdk import __version__ as __lakefs_sdk_version__
+from lakefs_sdk.models import Pagination
 
 lakefs_sdk_version = tuple(int(v) for v in __lakefs_sdk_version__.split("."))
 del __lakefs_sdk_version__
 
+T = TypeVar("T")
+P = ParamSpec("P")
+
+
+class PaginatedApiResponse(Protocol[T]):
+    pagination: Pagination
+    results: list[T]
+
+
+def depaginate(
+    api: Callable[P, PaginatedApiResponse[T]], *args: P.args, **kwargs: P.kwargs
```

Comment:
I'm not sure the type hints really do us any good here. How about just using `cast(depaginate(), ...)` on the results in the callers? E.g.,

```python
objects = list(depaginate(self.client.objects_api.list_objects, repository, ref, **kwargs))
for obj in cast(list[ObjectStats], objects):
    # obj is of type ObjectStats
    ...
```

Reply:
Sounds good, also solves the 3.9 problems with using `ParamSpec`.

I don't think mypy can even do anything with the spec, I tried inserting a random keyword argument into the `list_objects` call and it didn't throw any error.

Reply:
Taken over in [5a5a1dd](https://github.com/aai-institute/lakefs-spec/pull/180/commits/5a5a1ddb15a51c411baaecb97e3bc47266da933c).

Reply:
One minor thing, I believe the cast should be (if I'm not mistaken):

```python
objects = depaginate(self.client.objects_api.list_objects, repository, ref, **kwargs)
for obj in cast(Iterable[ObjectStats], objects):
```

Note the `Iterable`, since the generator doesn't return a `list`.

Reply:
Fixed in [a6661d9](https://github.com/aai-institute/lakefs-spec/pull/180/commits/a6661d99e25552041635d576ca0c67caf56e2f21).